### ISSUE TITLE
URI::Params#merge! and URI#merge_params!

### ISSUE THREAD

**Issue by skinnyjames** [crystal-lang/crystal#13399]
#### Problem statement

From a [discussion on the official chat](https://matrix.to/#/!DCnOFYRfeFHMcbUHoI:gitter.im/$7qbx7rGnBjOkx-i2vH8fz6FB-iAASnrDNDB0HMYOop4?via=gitter.im&via=matrix.org&via=nitro.chat)

> if you have an url (in string form) like this /path/to/resource?with_some_args=yes&another_one=also and you want to add more query parameters to it in a safe way, you can't easily

I imagine one use case might be to make a modify an existing URI to mutate / assign pagination parameters.

```crystal
require "uri"

uri = URI.parse("/resource?page=1&per_page=10")
params = uri.query_params
params["page"] = "2"
uri.query_params = params

puts uri.to_s
```
 
 to 
 
 ```crystal
uri = URI.parse("/resource?page=1&per_page=10")
         .merge_params!(URI::Params.new({ "page" => ["2"]}))

puts uri.to_s
```

#### Alternatives

Presented above.  `URI#query_params` isn't a [stateful method](https://github.com/crystal-lang/crystal/blob/016578f85/src/uri.cr#L270), so proposing 2 methods to mutate both `URI::Params` and `URI` objects.

#### Proposal

2 new methods: example here: https://play.crystal-lang.org/#/r/ey0d

An idea for supporting mutating query params, but also allowing overriding or appending existing  param values with other param values on a `URI::Params` object.

#### Alternative solutions

Since the proposal uses the `reduce` block to keep track of existing param state while performing a side effect, there may be a better solution, but I'd typically use recursion for that, which probably isn't appropriate.

#### Backward compat

Shouldn't be an issue, as the proposal is only new methods which use the current interface for appending and replacing params.

**Comment by straight-shoota** [comment]
Sounds good.

I suppose an alternative could be to turn `URI#query_params` into a stateful method and store the parsed params in `URI`. This might have other side effects, though.

**Comment by skinnyjames** [comment]
> Sounds good.
> 
> I suppose an alternative could be to turn `URI#query_params` into a stateful method and store the parsed params in `URI`. This might have other side effects, though.

Thanks!  I Thought more about it via conversation, and It seems more advantageous to be able to be able to operate on `URI::Params` and also commit them back to the URI object, which seems to already being storing the query string. 

I'd like to refactor the proposal to have `URI::Params#merge!` and `URI#query_params(&)`.  The latter converts the query string to params, yields it, and commits it back to the query string.   Example here: https://play.crystal-lang.org/#/r/ey31

This would probably be a more robust method that's still backward compatible.   Happy to stand up an PR if it seems like a good add.

**Comment by skinnyjames** [comment]
Putting together a PR for this with specs now.

**Comment by straight-shoota** [comment]
I think it would be good to express the mutating nature of the method in its name: `URI#update_query_params`.

**Comment by HertzDevil** [comment]
I don't think `#update_query_params` should return a newly parsed instance of the query parameters, because the caller doesn't always need it. Instead methods like this often return `self` to enable chaining

**Comment by skinnyjames** [comment]
> I don't think #update_query_params should return a newly parsed instance of the query parameters, because the caller doesn't always need it. Instead methods like this often return self to enable chaining

That's right.  it was originally just `#query_params(&) : URI::Params`.  That makes even less sense now.

However, the [Interpreter tests are failing](https://github.com/crystal-lang/crystal/actions/runs/4981251313/jobs/8915740543#step:6:759) with the new implementation, and I don't have the time to look into why.
I'm happy to to revert the implementation to using reduce, or hand off the PR.
